galley/pkg/crd/equals.go

```diff
@@ -0,0 +1,35 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package crd
+
+import (
+	"reflect"
+
+	ext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+
+	"istio.io/istio/galley/pkg/common"
+	"istio.io/istio/pkg/log"
+)
+
+// Equals checks whether the given two CRDs are equal or not.
+func equals(c1 *ext.CustomResourceDefinition, c2 *ext.CustomResourceDefinition) bool {
```

Comment:
What is the argument against reflect.DeepEqual? I know it would be slow due to reflection, but at the scale (and level) we're doing things it is a minuscule cost compared to the other costs in the system. It also uses an already available method without incurring external dependencies.

I looked at hashutil.DeepHashObject, which seems like a fine way of doing the same thing, but it looks like we will need to add a dependency for it. 

bytestring equal has the downside of having to introduce an error path, as json.Marshal will return an error. Not to mention the additional overhead of additional heap allocations, sorting and transformations.

Unless there is a fundamental, semantic problem that I am missing with using reflect.DeepEqual, I'd prefer to keep it.

Reply:
hashutil.DeepHashObject uses reflect indirectly via https://github.com/davecgh/go-spew.

https://godoc.org/github.com/google/go-cmp/cmp is another option.

Reply:
consistency is one argument. We'd better to use the same mechanism to do the same task as k8s. it will save our testing overhead.